### ISSUE TITLE
ZOD frames data format documentation

### ISSUE THREAD

**Issue by nlgranger** [zenseact/zod#59]
Unless I completely missed it, I don't believe the data format is documented.

A few things were not obvious to me, and I'd like confirmation about it:

- *Timestamp of annotations*: timestamp of the lidar frame closest to image timestamp ?
- *Coordinates system for annotations*: lidar of the frame closest to image timestamp ?
- *Ego movement compensation*: the lidar point cloud is not compensated for ego car movement?
- *Lidar timestamp*: if the point cloud is not compensated, what is the meaning of the point cloud timestamp ?

**Comment by Petros626** [comment]
Coordinate system is probably camera frame like in KITTI, but in a issue Lidar seems to be the reference. 

@atonderski can you guys please add these missing information? Are calibration files provided for frame transformation?

**Comment by wljungbergh** [comment]
Hi, I hope I can clarify some confusion.


**Timestamp of annotations:**
The annotations are made in the closest lidar frame indeed. This is the "key-frame".

**Coordinates system for annotations:**
The annotations are made in the LiDAR coordinate system. See our helper annotation classes for inspiration.

**Ego movement compensation:**
For all LiDAR data in the ZOD Frames, the point cloud has been compensated using the ego motion to the LiDAR timestamp.
For all LiDAR data in the ZOD Sequences, the point cloud has not been compensated. 

**Lidar timestamp:**
While not strictly necessary, the timestamp in the uncompensated point cloud sets a common reference between sweeps over time. This timestamp reflects when the LiDAR is at a certain angle. If one would want to do compensation of these point clouds, one would transform all points to this timestamp and get the compensation-discontinuity at the same angle for all sweeps, which is generally good.


@Petros626, there are calibration files between all sensors in the data.

**Comment by Petros626** [comment]
@wljungbergh thank you for the quick answer. 
What was the specific reason you used the lidar frame for annotation reference and not camera? 

update: Why the `LiDARs` axes points to the right direction of the vehicle instead to the front of the car like the `OxTS`?

**Comment by nlgranger** [comment]
Well it's taken me forever to find time to work on this again. Anyhow, thanks for the detailed answer @wljungbergh.

I was able to integrate ZOD Frames to my dataset library: https://cea-list.github.io/tri3d/generated/tri3d.datasets.ZODFrames.html

**Comment by Petros626** [comment]
@nlgranger did you provide lidar to camera boxes transformation?

**Comment by nlgranger** [comment]
You should be able to get the transforms between all sensors/timestamps.

As for the boxes, you can request them in any the coordinates of any sensor and frame, tri3d will interpolate from boxes at the nearest timestamp(s). Since ZOD Frames only has one annotated (lidar) frame, the interpolation doesn't really work outside of the annotated lidar frame and the nearest camera frame.

**Comment by Petros626** [comment]
What exactly is interpolation all about in your custom implementation? Generally only the 3D bboxes in der LiDAR frame are in my interest, so does the interpolation add additional 3D bboxes in the scene?

**Comment by nlgranger** [comment]
I should probably document the inner working of `.boxes()` more, now that it is common to all datasets. Basically it works like this:
1. [Locate the timestamp](https://github.com/CEA-LIST/tri3d/blob/v0.2.1/tri3d/datasets/dataset.py#L605-L608) of the requested frame in the box annotation timeline
2. [Interpolate each box tracklet](https://github.com/CEA-LIST/tri3d/blob/v0.2.1/tri3d/datasets/dataset.py#L652-L677) between the previous and following timestamps (if the time difference to one of the timestamps is very small, optimize by [returning boxes returned as-is](https://github.com/CEA-LIST/tri3d/blob/v0.2.1/tri3d/datasets/dataset.py#L610-L638))
3. Transform the boxes from their original coordinates to the requested one.

The figure on the README illustrates the process:
![box interpolation](https://github.com/CEA-LIST/tri3d/raw/v0.2.1/docs/source/timelines.svg)

**Comment by Petros626** [comment]
Hey @nlgranger,

thanks for your effort!

I understand the principle behind it, but the ZOD dataset is completely labeled with 3D annotations and corresponding images to the LiDAR frame or not. There are annotated images (keyframes) with matching LiDAR point clouds. These are synchronised ?!
→ For each keyframe, there is an image, a .npy (LiDAR), and the annotations match each other.

You're referring to the sequences, right?
For normal ZOD sequences, you cannot do a 1:1 mapping (unless you interpolate yourself and accept that there are no perfect annotations). Am I right?

**Comment by nlgranger** [comment]
> For normal ZOD sequences, you cannot do a 1:1 mapping (unless you interpolate yourself and accept that there are no perfect annotations). Am I right?

Yes, but only ZOD **Frames** is supported in tri3d. Since context frames data is provided as well as the keyframe, I consider that as a sequence.
But only one keyframe per scene is annotated, so tri3d is unable to extrapolate the trajectories. I mean it will return them at whatever context frame you ask, but the boxes will be off for moving objects (and for static ones too because the ego pose data is a bit noisy).

**Comment by Petros626** [comment]
I understand for my use case with the frames it's totally fine. You were searching for a solution with the interpolation approach, because the annotations for the sequences are still not released?

**Comment by nlgranger** [comment]
Honestly I just wanted to inspect the data and figured it was easy enough to support it in the library. I'm not using ZOD at the moment.

**Comment by Petros626** [comment]
@nlgranger Hey, hope your doing well,

I got a theoretical question from my master thesis and don't know how to solve it now. In general I use `zod` with `OpenPCDet` and came across a coordinate transformation (alignment from `zod` to `kitti`) problem. We transformed the `zod` LiDAR frame into the `kitti` frame and shifted the data origin height of the sensor (~2.08702246m -> 1.73m) too. The consideration is that when I work with functions of the ZOD API, it expects the LiDAR data in the ZOD frame and not in a differently transformed frame. Only in the further processing can the data be used as KITTI frame, right?

Template as orientation:
get the LiDAR data (transformed)
https://github.com/griesbchr/3DTrans/blob/3174699105aefb3ed11e524606f707fd91239850/pcdet/datasets/zod/zod_dataset.py#L240

get information from zod
https://github.com/griesbchr/3DTrans/blob/3174699105aefb3ed11e524606f707fd91239850/pcdet/datasets/zod/zod_dataset.py#L670

This user directly uses his transformed data (Waymo Frame) and calculates the required parameter. The origin comes from https://github.com/griesbchr/3DTrans/blob/3174699105aefb3ed11e524606f707fd91239850/pcdet/datasets/kitti/kitti_dataset.py#L289

Remember:
`zod` is in LiDAR frame, annotations too
`kitti` is in Camera frame, annotations too.

I'd really appreciate any valueful discussion and support!

**Comment by nlgranger** [comment]
Hi and sorry for the delay. If I understand correctly, your have to deal with two conventions. In my experience for short(ish) term projects with one or two datasets, it's safer to adhere to whatever convention is used by your model and evaluation code. For long-term projects you standardize the datasets to a common ground.

**Comment by Petros626** [comment]
Hey no problem,

I experimented myself and find a way to adapt the coordinate system from ZOD to KITTI for my purposes. 

Btw did you experienced the same (https://github.com/zenseact/zod/issues/80, https://github.com/zenseact/zod/issues/81) with ZOD?
